tensorflow/compiler/mlir/tosa/transforms/legalize_tfl_stateful.cc

```diff
@@ -0,0 +1,208 @@
+    DenseMap<StringRef, LocationAttr> sharedNameToLoc;
+
+    llvm::SmallVector<mlir::TFL::VarHandleOp, 6> handleOps;
+    llvm::SmallVector<mlir::TFL::AssignVariableOp, 6> assignOps;
+    llvm::SmallVector<mlir::TFL::ReadVariableOp, 6> readOps;
+
+    for (auto it : symNameToFunction) {
+      auto func = std::get<1>(it);
+      // Look through the initialization functions and find the assigned values
+      // for each handle, save out the constant value.
+      for (auto init : func.getOps<mlir::TFL::CallOnceOp>()) {
+        auto findInitFunc = symNameToFunction.find(init.getSessionInitFunctionAttr());
+        if (findInitFunc == symNameToFunction.end()) {
+          init.emitError("Unable to find initialization function: ");
+          continue;
+        }
+        func::FuncOp initFunc = std::get<1>(*findInitFunc);
+        for (auto assign : initFunc.getOps<mlir::TFL::AssignVariableOp>()) {
+          // 1. var_handle part
+          auto handle = dyn_cast<mlir::TFL::VarHandleOp>(assign.getResourceId().getDefiningOp());
+          if (!handle) continue;
+
+          // 2. pseudo_const part
+          DenseElementsAttr constant;
+          if (!matchPattern(assign.getValue(), m_Constant(&constant))) {
+            // Quantized types we can not use the m_Constant matcher.
+            if (auto constOp = dyn_cast<mlir::TFL::QConstOp>(
+                    assign.getValue().getDefiningOp())) {
+              constant = constOp.getValue().cast<DenseElementsAttr>();
+            }
+          }
+          if (!constant) continue;
+
+          auto name = handle.getSharedName();
+          sharedNameToConstant[name] = constant;
+          sharedNameToLoc[name] = handle.getLoc();
+        }
+      }
+
+      // We also want to grab the list of operations to replace.
+      for (auto& op : func.getOps()) {
+        if (auto handle = dyn_cast<mlir::TFL::VarHandleOp>(op))
+          handleOps.push_back(handle);
+        if (auto assign = dyn_cast<mlir::TFL::AssignVariableOp>(op))
+          assignOps.push_back(assign);
+        if (auto read = dyn_cast<mlir::TFL::ReadVariableOp>(op))
+          readOps.push_back(read);
+      }
+    }
+
+    // TF::CallOnceOps are no longer needed as we have already extracted their
+    // state.
+    SmallVector<mlir::TFL::CallOnceOp> callOnceOps;
+    for (auto func : moduleOp.getOps<func::FuncOp>()) {
+      for (auto init : func.getOps<mlir::TFL::CallOnceOp>()) {
+        callOnceOps.push_back(init);
+      }
+    }
+    for (auto op : callOnceOps) op.erase();
+
+    // Create the Tosa_VariableOps to store our new global variables.
+    DenseMap<StringRef, mlir::tosa::VariableOp> symbolRefMap;
+    for (auto it : sharedNameToConstant) {
+      auto name = std::get<0>(it);
+      auto attribute = std::get<1>(it);
+      auto locIt = sharedNameToLoc.find(name);
+      LocationAttr loc = mlir::UnknownLoc();
+      if (locIt != sharedNameToLoc.end()) {
+        loc = std::get<1>(*locIt);
```

Comment:
This should probably be handled while creating `sharedNameToLoc`

Reply:
done.